### PR TITLE
Merge Hyperion into MX-Bluesky

### LINKED ISSUE #169: Merge Hyperion into this repo
We decided today that having the two `mx-bluesky` and `hyperion` repos is causing us too much dependency overhead. Hyperion should be merged into here.

Things that must happen when they are merged:
* Git history of hyperion is preserved
* All hyperion tests are still run and pass
* Issues on hyperion repo are moved to mx-bluesky (with an appropriate label)

Things that should happen (e.g. make a new issue for them if hard):
* The structure of `mx-bluesky` needs to be well documented
* All the GH actions that move things on the Hyperion project board still work (with the caveat that they only work on things tagged with the hyperion label)
* Slack messages on releases still get pushed to the `hyperion` channel

Things we would then like to add:
* Release notes are now going to be a bit messy, it would be nice to have an automated way of adding a `hyperion` heading to them

Nice to haves:
* Navigating to https://github.com/DiamondLightSource/hyperion redirects to https://github.com/DiamondLightSource/mx-bluesky

### PR DESCRIPTION
**Pull request by dperl-dls**
Brings all the code from hyperion into mx-bluesky to reduce our dependency problems

Fixes #169 

### Done:
- Updates all commit messages copied from hyperion to link to their original issues
- Fixes or marks all the type checking issues which were in hyperion (now with both `pyright` and `mypy`!)
- Tests should all pass
- Drops support for python 3.10, policy for review/comment here: https://github.com/DiamondLightSource/mx-bluesky/blob/merge_in_hyperion/docs/developer/general/explanations/decisions/0003-python-version-support.rst
- Rearranges the directory structure somewhat, policy for review/comment here:
https://github.com/DiamondLightSource/mx-bluesky/blob/merge_in_hyperion/docs/developer/general/explanations/decisions/0002-repository-structure.rst
- Original hyperion workflows are copied over to `hyperion_other/workflows` for adapting to this repo

### TODO:

- [ ] update the release workflow to work with pyproject.toml
- [ ] update the deploy workflow to work with pyproject.toml
- [ ] fix codecov integration
- [ ] get issue/project integration to work
- [ ] rework hyperion's readme & the wiki docs to the mx-bluesky sphinx docs structure
- [ ] move hyperion issues here and tag them
- [ ] update slack integration
- [ ] add a linting step to ensure clean import directionality

### CODE REVIEW

**Review by noemifrisina — COMMENTED** [review]
Looks like a good start, thank you! 

I think it's fine to drop python3.10, asaik the other applications are using 3.11 anyway so don't see the point in getting the hyperion code to work with 3.10.

It would be good to agree on the priority of the TODOs and open issues for them. Also I'd add this to the list: https://github.com/DiamondLightSource/mx-bluesky/blob/86dc96ca97f204e26cf43a5bad9420d560a08ff7/docs/developer/general/explanations/decisions/0002-repository-structure.rst?plain=1#L44

**Comment by olliesilvester** [comment]
The commits here are the unsquashed versions of Hyperion. Not a huge deal though, as we only recently moved to using squashed commits in Hyperion

**Review by olliesilvester — CHANGES_REQUESTED** [review]
I would prefer if the commit history was preserved to the squashed version, but if that's too much effort then I'm happy to approve, since this is currently blocking any more Hyperion development. We should create new tickets for the remaining things to do as well, and try out deployment fairly quickly

**Comment by dperl-dls** [comment]
It does have the squashed versions of the ones which were squashed

**Review by olliesilvester — APPROVED** [review]

**Review by noemifrisina — APPROVED** [review]

File: pyproject.toml [hunk 1]
```diff
@@ -34,10 +56,17 @@ requires-python = ">=3.10"
 [project.optional-dependencies]
 dev = [
     "black",
+    "black",
+    "build",
     "copier",
+    "diff-cover",
     "GitPython",
+    "ipython",
+    "matplotlib",
```
Comment (rtuck99) [inline pyproject.toml:65]: matplotlib is no longer optional

File: pyproject.toml [hunk 2]
```diff
@@ -34,10 +56,17 @@ requires-python = ">=3.10"
...
+    "mockito",
     "mypy",
     "myst-parser",
+    "ophyd-async",
```
Comment (rtuck99) [inline pyproject.toml:69]: ditto

File: pyproject.toml [hunk 3]
```diff
@@ -34,10 +56,17 @@ requires-python = ">=3.10"
 [project.optional-dependencies]
 dev = [
     "black",
+    "black",
```
Comment (rtuck99) [inline pyproject.toml:59]: duplicate line

File: pyproject.toml [hunk 4]
```diff
@@ -7,25 +7,47 @@ name = "mx-bluesky"
...
+    "blueapi == 0.4.5a1",
+    "daq-config-server >= 0.1.1",
+    "ophyd == 1.9.0",
+    "ophyd-async >= 0.3a5",
+    "bluesky >= 1.13.0a4",
     "dls-dodal @ git+https://github.com/DiamondLightSource/dodal.git",
-    "blueapi==0.4.5-a1",
 ]
+
+
```
Comment (rtuck99) [inline pyproject.toml:50]: requires-python should be 3.11

**Review by rtuck99 — APPROVED** [review]
Approved, please fix minor things with `pyproject.toml`

**Pull request merged**